for_dict_challenges_bonus.py

```diff
@@ -55,5 +55,136 @@ def generate_chat_history():
+def first_issue(user_messages):
+    wrote_the_most = max(user_messages.values(), key=len)
+    for _ in user_messages:
+        if user_messages[_] == wrote_the_most:
+            wrote_the_most = _
+    return wrote_the_most
+
+
+def number_answers(messages):
+    number_answers = {}
+    for _ in messages:
+        if _['reply_for'] != None and _['reply_for'] != []:
+            if _['reply_for'] in number_answers:
+                number_answers[_['reply_for']] += 1
+            else:
+                number_answers[_['reply_for']] = 1
+    return number_answers
+def second_issue(number_answers):
+    number_responses_per_user = {}
+    for _ in chat_history:
+        message_id = _['id']
+        for i in number_answers:
+            if i == message_id:
+                if number_responses_per_user.get(_['sent_by']) == None:
+                    number_responses_per_user[_['sent_by']] = number_answers[i]
+                else:
+                    number_responses_per_user[_['sent_by']] += number_answers[i]
+    most_cited = max(number_responses_per_user.values())
+    for _ in number_responses_per_user:
+        if number_responses_per_user[_] == most_cited:
+            most_cited = _
+    return most_cited
+
+
+def how_many_views(user_messages):
+    how_many_views = {}
+    reply = []
+    for _ in user_messages:
+        who_saw = []
+        message_list = user_messages[_]
+        for i in message_list:
+            for j in chat_history:
+                if j['id'] == i:
+                    who_saw += j['seen_by']
+        how_many_views[_] = len(who_saw)
+    most_viewed = max(how_many_views.values())
+    for _ in how_many_views:
+        if how_many_views[_] == most_viewed:
+            reply.append(_)
+    return reply
+
+
+def sending_time(messages):
+    sending_time = {
+        'morning': 0,
+        'day': 0,
+        'evening': 0,
+        'night': 0
+    }
+    for _ in messages:
+        time = int(_['sent_at'].strftime('%H'))
+        if 6 <= time < 12:
+            sending_time['morning'] += 1
+        elif 12 <= time < 18:
+            sending_time['day'] += 1
+        elif 18 <= time < 23:
+            sending_time['evening'] += 1
+        else:
+            sending_time['night'] += 1
+    busiest_time = max(sending_time.values())
+    for _ in sending_time:
+        if sending_time[_] == busiest_time:
+            return _
+
+
+def is_there_answer(id):
+    for _ in chat_history:
+        if _['reply_for'] == id:
+            return True
+    return False
+def reply_for_from_id(id):
+    for _ in chat_history:
+        if _['id'] == id:
+            return _['reply_for']
+def message_threads(messages):
+    message_threads = {}
+    reply = []
+    for _ in messages:
+        answer = is_there_answer(_['id'])
+        count = 0
+        by = _['id']
+        beginning = _['reply_for']
+        if answer == False:
+            while beginning != None and beginning != []:
```

Comment:
is not None
любой не пустой словарь будет true, можно сделать
```suggestion
            while beginning is not None and beginning:
```